### PR TITLE
Add request recording

### LINKED ISSUE #168: Java supports requests recording

### PR DESCRIPTION
**Pull request by apotterri**
By default, a request made to a Spring Boot app will now generate a request recording.

Fixes #168 .

### CODE REVIEW

File: agent/src/main/java/com/appland/appmap/transform/ClassFileTransformer.java [hunk 1]
```diff
@@ -176,6 +193,10 @@ public byte[] transform(ClassLoader loader,
           continue;
         }
 
+        if (behavior.getName().equals("doFilter")) {
```
Comment (apotterri) [inline agent/src/main/java/com/appland/appmap/transform/ClassFileTransformer.java:196]: This debugging was removed in a subsequent commit.

**Review by dividedmind — APPROVED** [review]
I had a couple of questions, but my Java is a bit rusty so please forgive me if they're trivial or misguided :) Other than that, it looks ok! Hopefully hooking other frameworks won't be too difficult.

File: agent/src/main/java/com/appland/appmap/util/AppMapTinylogConfigurationLoader.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        try (InputStream stream = cl.getResourceAsStream(configFile)) {
+          if (stream != null) {
+            properties.load(stream);
+          }
+        }
+      }
+    }
+
+    try {
+      if (localConfigFile.exists()) {
```
[the remark this answers is not in the document]
Comment (apotterri) [reply agent/src/main/java/com/appland/appmap/util/AppMapTinylogConfigurationLoader.java:36]: Having a local config file is optional, so I think this check makes sense. If the file gets removed after the check, logging an error seems ok.

File: agent/src/main/java/com/appland/appmap/config/AppMapConfig.java [hunk 3]
```diff
@@ -237,4 +243,24 @@ public void accept(Path packagePath) {
     }
     return sw.toString();
   }
+
+  public static TaggedLogger configureLogging() {
+    // Make sure the app's instance of tinylog (if it has one) doesn't install a
+    // shutdown hook. We'll take care of it.
+    System.setProperty("tinylog.autoshutdown", "false");
```
[the remark this answers is not in the document]
Comment (apotterri) [reply agent/src/main/java/com/appland/appmap/config/AppMapConfig.java]: Yes, you're right, good catch! 

I dug into this a little more. I was able to convince myself that shadowing makes this unnecessary. By renaming classes, it separates the state of the embedded tinylog from that of the app's instance. (In this case, the `ProviderRegistry` [here](https://github.com/getappmap/appmap-java/pull/183/files#diff-f21749602122689f6874d2a0c1811f06964b6cb87f516e580bd3d75306e1f75cR59) is distinct from the app's, so the app won't try to shut down our tinylog's writer thread.)

I removed this change.

File: agent/src/main/java/com/appland/appmap/process/hooks/http/ServletListener.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  }
+
+  private static void requestInitialized(ServletRequestEvent evt) {
+    HttpServletRequest req = evt.getServletRequest();
+    ServletContext ctx = req.getServletContext();
+    String ctxName = ctx.getServletContextName();
+    String uri = req.getRequestURI();
+    logger.trace("{}: request init, {} {}", ctxName, req.getMethod(), uri);
+
+    if (uri.equals("/_appmap/record")) {
```
[the remark this answers is not in the document]
Comment (apotterri) [reply agent/src/main/java/com/appland/appmap/process/hooks/http/ServletListener.java:45]: It's a misnomer. This "uri" here is just the path, and this check ensures that requests for `/_appmap/record` don't appear in the recording.

**Pull request merged**